pkg/backends/calico/client.go

```diff
@@ -811,6 +885,258 @@ func (c *client) onUpdates(updates []api.Update) {
 	c.onNewUpdates()
 }
 
+func (c *client) updateBGPConfigCache(update api.Update) {
+	key, ok := update.KVPair.Key.(model.ResourceKey)
+	if !ok {
+		log.Infof("Error getting BGP Configuration key. %#v",ok)
+	}
+
+	resourceName := key.Name
+	v3res, ok := update.KVPair.Value.(*apiv3.BGPConfiguration)
+	if !ok {
+		log.Infof("BGP Configuration is either empty or error occurred while getting BGPConfiguration value %#v", ok)
+		v3res = nil
+	}
+
+	if resourceName == globalConfigName {
+		c.getPrefixAdvertisementsKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getListenPortKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getASNumberKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getServiceExternalIPsKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getServiceClusterIPsKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getNodeToNodeMeshKVPair(v3res, model.GlobalBGPConfigKey{})
+		c.getLogSeverityKVPair(v3res, model.GlobalBGPConfigKey{})
+	} else {
+		// update node-specific BGPConfig
+		var nodeName string
+		if strings.HasPrefix(resourceName, perNodeConfigNamePrefix) {
+			nodeName = resourceName[len(perNodeConfigNamePrefix):]
+		}
+		c.getPrefixAdvertisementsKVPair(v3res, model.NodeBGPConfigKey{Nodename: nodeName})
+		c.getListenPortKVPair(v3res, model.NodeBGPConfigKey{Nodename: nodeName})
+		c.getLogSeverityKVPair(v3res, model.NodeBGPConfigKey{Nodename: nodeName})
+	}
+}
+
+func getBGPConfigKey(v1KeyName string, key interface{}) model.Key {
+	switch k := key.(type) {
+	case model.NodeBGPConfigKey:
+		k.Name = v1KeyName
+		return &k
+	case model.GlobalBGPConfigKey:
+		k.Name = v1KeyName
+		return &k
+	default:
+		log.Warning("Bad value for BGP Configuration key.")
+		return nil
+	}
+}
+
+// Returns a model.KVPair for the given key and value. If no value is provided, returns model.KVPair with key.
+func getKVPair(key model.Key, value ...string) *model.KVPair {
+	if len(value) > 0 {
+		return &model.KVPair{
+			Key:   key,
+			Value: value[0],
+		}
+	}
+	return &model.KVPair{
+		Key: key,
+	}
+}
+
+func (c *client) getPrefixAdvertisementsKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKeyIPv4 := getBGPConfigKey("prefix_advertisements/ip_v4", key)
+	globalKeyIPv6 := getBGPConfigKey("prefix_advertisements/ip_v6", key)
```

Comment:
nit: I think `global` is misleading here, as it might be a node-specific key.  Is that right?

Reply:
If so, I suggest removing "global" from all of these functions, as they all follow a signature that allows them to be used for both global and node-specific.